api/src/opentrons/hardware_control/emulation/parser.py

```
@@ -0,0 +1,75 @@
+import re
+from dataclasses import dataclass
+from typing import Sequence, Dict, Generator, Optional
+
+
+@dataclass
+class Command:
+    gcode: str
+    body: str
+    params: Dict[str, Optional[float]]
+
+
+class Parser:
+    """Gcode line parser."""
+
+    PREFIX_NUMBER_RE = re.compile(r"(?P<prefix>[A-Z])(?P<value>-?\d*\.?\d*)")
+
+    def __init__(self, gcodes: Sequence[str]) -> None:
+        """
+        Construct a parser.
+
+        Args:
+            gcodes: a sequence of gcodes this parser supports.
+        """
+        # Sort in reversed size order to find most specific first.
+        ordered = reversed(sorted(gcodes, key=lambda x: len(x)))
+        # Create a regular expression that looks like this
+        #  For valid gcodes: G1, G123, G123.2 ->
+        #       (G123.2|G123|G1)(?!\d)
+        #  Match any of the gcodes not followed by a digit.
+        self._re = re.compile(r'({})(?!\d)'.format('|'.join(ordered)))
+
+    def parse(self, line: str) -> Generator[Command, None, None]:
+        """
+        Parse a line to extract commands.
+
+        Args:
+            line: a line containing gcodes and their values. spaces are optional.
+
+        Returns:
+            Command object
+        """
+        previous = None
+        for i in self._re.finditer(line):
```

I haven't tested this, but I *think* this will silently discard stuff at the end of the line that we can't parse, and possibly also stuff in the middle of the line that we can't parse.

I think we should error in those cases, instead.

---

Yes. It only finds matches and ignores the rest.

---

This was a great insight, @SyntaxColoring . Revealed a potential crummy bug.